### ISSUE TITLE
Should the operators `#` `by` `align` be promotable?

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#19348]
Currently the operators `#` `by` `align` work on a range or rectangular domains and a suitable index. `#` also works on rectangular arrays.

What should the behavior be when they are invoked on something else? The following two policies are easier to implement:
* (S) Issue a "specialized" error clearly stating that this use is currently not supported.
* (U) Treat them like a user-defined function would be treated. Namely:
  - attempt to apply promotion, then
  - issue a "generic" error saying "unresolved call, there are NN candidates"

What behavior do we want - (S) or (U) or something else?

I see this as a tradeoff between:
* elegance -- fewer special cases, more uniform availability of promotion, vs.
* ease of understanding the program behavior -- promotion can be subtle and unexpected to beginners.

I am raising this question because our implementation looks inconsistent. For example:
```chpl
var arrayOfRanges: [1..11] range = 1..9;
iter iterOfRanges() { yield 1..9; yield 11..19; }

writeln(6 # 7);               // generic error
writeln(arrayOfRanges # 7);   // array #, prints 1..9 1..9 1..9 1..9 1..9 1..9 1..9
writeln(iterOfRanges() # 7);  // promoted #, prints 1..7 11..17

writeln(6 by 7);              // specific error
writeln(arrayOfRanges by 7);  // specific error
writeln(iterOfRanges() by 7); // specific error
```

The current implementation is more consistent in disallowing promotion over the second argument and reporting a specific error instead, which may or may not be the desired behavior.

This question may apply to other operators as well, ex. #19254, #19264, #19332. Somewhat related: #19261.

**Comment by bradcray** [comment]
I agree that these should be treated more consistently with one another.  What does the overload look like that results in the errors for the `by` operator?  And what does the error look like?

**Comment by vasslitvinov** [comment]
The overload is:
```chpl
operator by(r, step) {...}
```
The error message is as follows, which can be improved:
```
the first argument of the 'by' operator is not a range
```

If we mark this overload with pragma "last resort", we will get the option (PFS).
Currently resolution will not try forwarding before trying "last resort", however arguably it should.

**Comment by bradcray** [comment]
I'd mark it (and, more generally, any error overloads that we come across) as "last resort".

**Comment by lydia-duncan** [comment]
@vasslitvinov - did #19356 resolve this issue?  Or is there more discussion/implementation work still needed?

**Comment by vasslitvinov** [comment]
> did https://github.com/chapel-lang/chapel/pull/19356 resolve this issue? Or is there more discussion/implementation work still needed?

Good question. I am not sure about the answer.

For now, there are no plans to make any changes for 2.0 w.r.t. this issue, so I removed the 2.0 tag.
